core/services/relay/evm/chain_reader.go

```diff
@@ -83,6 +88,79 @@ func (cr *chainReader) Bind(ctx context.Context, bindings []commontypes.BoundCon
 	return cr.contractBindings.Bind(ctx, bindings)
 }
 
+func (cr *chainReader) QueryKey(ctx context.Context, key string, queryFilter commontypes.QueryFilter, limitAndSort commontypes.LimitAndSort, sequenceDataType any) ([]commontypes.Sequence, error) {
+	remappedQueryFilter, err := remapQueryKeyFilters(key, cr.eventIndexBindings, queryFilter)
+	if err != nil {
+		return nil, err
+	}
+
+	logs, err := cr.lp.FilteredLogs(remappedQueryFilter, limitAndSort)
+	if err != nil {
+		return nil, err
+	}
+
+	return cr.eventIndexBindings.DecodeLogsIntoSequences(ctx, key, logs, sequenceDataType)
+}
+
+func (cr *chainReader) QueryKeys(ctx context.Context, keys []string, queryFilter commontypes.QueryFilter, limitAndSort commontypes.LimitAndSort, sequencesDataTypes []any) ([][]commontypes.Sequence, error) {
+	if len(keys) != len(sequencesDataTypes) {
+		return nil, fmt.Errorf("length of keys and sequencesDataTypes must be the same")
+	}
+
+	var sequencesMatrix [][]commontypes.Sequence
+	for i, key := range keys {
+		sequences, err := cr.QueryKey(ctx, key, queryFilter, limitAndSort, sequencesDataTypes[i])
+		if err != nil {
+			return nil, err
+		}
+		sequencesMatrix = append(sequencesMatrix, sequences)
+	}
+
+	return sequencesMatrix, nil
+}
+
+func (cr *chainReader) QueryByKeyValuesIn(ctx context.Context, keyValueIn commontypes.KeyValueIn, queryFilter commontypes.QueryFilter, limitAndSort commontypes.LimitAndSort, sequenceDataType any) ([]commontypes.Sequence, error) {
+	remappedQueryFilter, err := remapQueryKeyByValuesFilters(keyValueIn, cr.eventIndexBindings, queryFilter)
+	if err != nil {
+		return nil, err
+	}
+
+	logs, err := cr.lp.FilteredLogs(remappedQueryFilter, limitAndSort)
+	if err != nil {
+		return nil, err
+	}
+
+	return cr.eventIndexBindings.DecodeLogsIntoSequences(ctx, keyValueIn.Key, logs, sequenceDataType)
+}
+
+// TODO values shouldn't be string
+func (cr *chainReader) QueryByKeysValuesIn(ctx context.Context, keysValuesIn []commontypes.KeyValueIn, queryFilter commontypes.QueryFilter, limitAndSort commontypes.LimitAndSort, sequencesDataTypes []any) ([][]commontypes.Sequence, error) {
+	if len(keysValuesIn) != len(sequencesDataTypes) {
+		return nil, fmt.Errorf("length of keysValuesIn sequencesDataTypes must be the same")
+	}
+
+	var sequencesMatrix [][]commontypes.Sequence
+	for i, keyValuesIn := range keysValuesIn {
+		sequences, err := cr.QueryByKeyValuesIn(ctx, keyValuesIn, queryFilter, limitAndSort, sequencesDataTypes[i])
```

Comment:
Oh... I guess we're doing a loop here because LogPoller only supports matching on one key at a time? I think filtering on multiple keys was a feature we were thinking about adding at some point, I guess when we do hopefully we'll be able to call lp directly.

Reply:
Yeah, but also we have to fetch per sequence type(per key) since every the data has to be encoded/decoded differently

Reply:
> Yeah, but also we have to fetch per sequence type(per key) since every the data has to be encoded/decoded differently

Hmm... now I'm confused as that sounds very different from what I thought this was doing... maybe I'm not understanding the purpose of `QueryByKeysValuesIn`, I thought when you specify multiple keys you're just narrowing down the search, not changing anything about the type of data returned!  If it's affecting anything about the type of data returned, that would be a big change from the current way in which products fetch events from the chain.

For example, if you're wanting to search for transfers from wallet1 to wallet2, I was assuming you'd pass [`Key:"from", Value:"wallet1"` and `Key:"to", Value:"wallet2"]` for the list of key-value pairs, and then those would each filter down the number of events you get back... but all of the events should have the same data type and format... it shouldn't be returning some results with only one field filled in and other results with only another field filled in, all of them must have all the fields present in the same format. At least, if that's no longer going to be the case, that's a big change in what the products are expecting and I'm not sure how that would work.